src/modules/tools/laser/Laser.cpp

```diff
@@ -103,11 +106,77 @@ void Laser::on_module_loaded()
 
     //register for events
     this->register_for_event(ON_HALT);
+    this->register_for_event(ON_GCODE_RECEIVED);
+    this->register_for_event(ON_CONSOLE_LINE_RECEIVED);
 
     // no point in updating the power more than the PWM frequency, but no more than 1KHz
     THEKERNEL->slow_ticker->attach(std::min(1000UL, 1000000/period), this, &Laser::set_proportional_power);
 }
 
+void Laser::on_console_line_received( void *argument )
+{
+    if(THEKERNEL->is_halted()) return; // if in halted state ignore any commands
+
+    SerialMessage new_message = *static_cast<SerialMessage *>(argument);
+    string possible_command = new_message.message;
+
+    // ignore anything that is not lowercase or a letter
+    if(possible_command.empty() || !islower(possible_command[0]) || !isalpha(possible_command[0])) {
+        return;
+    }
+
+    string cmd = shift_parameter(possible_command);
+
+    // Act depending on command
+    if (cmd == "fire") {
+        string power = shift_parameter(possible_command);
+        if(power.empty()) {
+            new_message.stream->printf("Usage: fire power%%|off\n");
+        }
+        float p;
+        if(power == "off" || power == "0") {
+            p= 0;
+            new_message.stream->printf("turning laser off and returning to auto mode\n");
+
+        }else{
+            p= strtof(power.c_str(), NULL);
+            p= confine(p, 0.0F, 100.0F);
+            new_message.stream->printf("WARNING: Firing laser at %1.2f%% power, entering manual mode use fire off to return to auto mode\n", p);
+        }
+
+        p= p/100.0F;
+        manual_fire= true;
```

Comment:
This code looks almost the same as `set_proportional_power`.

Why not refactor common code into new `set_laser_power` and call it from here and `set_proportional_power`.

That way there is less duplicated code for manipulating laser pins.